lib/dao.js

```diff
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
+        if (doValidate === false) {
+          callConnector();
+        } else {
+          inst.isValid(function(valid) {
+            if (!valid) {
+              console.warn('Ignoring validation errors in updateOrCreate():');
+              console.warn('  %s', new ValidationError(inst).message);
+            }
+            callConnector();
+          }, update, options);
+        }
+        function callConnector() {
+          update = removeUndefined(update);
+          context = {
+            Model: Model,
+            where: ctx.where,
+            data: update,
+            currentInstance: inst,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('persist', context, function(err) {
+            if (err) return done(err);
+            connector.upsertWithWhere(modelName, where, update, done);
+          });
+        }
+        function done(err, data, info) {
+          if (err) return cb(err);
+          var context = {
+            Model: Model,
+            data: data,
+            isNewInstance: info && info.isNewInstance,
+            hookState: ctx.hookState,
+            options: options,
+          };
+          Model.notifyObserversOf('loaded', context, function(err) {
+            if (err) return cb(err);
+            var obj;
+            if (data && !(data instanceof Model)) {
+              inst._initProperties(data, { persisted: true });
+              obj = inst;
+            } else {
+              obj = data;
+            }
+            if (err) {
+              cb(err, obj);
+            } else {
+              var context = {
+                Model: Model,
+                instance: obj,
+                isNewInstance: info ? info.isNewInstance : undefined,
+                hookState: hookState,
+                options: options,
+              };
+              Model.notifyObserversOf('after save', context, function(err) {
+                cb(err, obj);
+              });
+            }
+          });
+        }
+      });
+    } else {
+      var element = Object.keys(where)[0];
+      var filter = { 'where': where };
+      if (!Object.keys(data).length) {
+        process.nextTick(function() {
+          cb('data field cannot be empty');
+        });
+        return cb.promise;
+      }
 
+      if (!data.hasOwnProperty(element)) {
+        data[element] = where[element];
+      }
+      var pk = idName(Model);
+      self.find(filter, function(err, instances) {
+        var modelsLength = instances.length;
+        if (modelsLength === 0 && data.hasOwnProperty(pk)) {
+          self.create(data, options, cb);
+          return cb.promise;
+        } else if (modelsLength == 0 && data.hasOwnProperty(pk) == false) {
+          cb('Error message: No matching instance found.Found error while trying ' +
+                'to create a new instance.' +
+                'Primary key and value ' +
+                'is not provided in the payload');
```

Comment:
@Amir-61 
For many connectors,it expects id field to be present in the data object while create operation is performed. In case of auto generated ids,  it expects {"id":""} to be present in the data object while create is performed and if not present the operation fails.  So, what would be the best thing to do here ?